src/main/java/org/jenkinsci/remoting/engine/JnlpConnectionState.java

```diff
@@ -0,0 +1,531 @@
+     *                               {@link JnlpConnectionStateListener#beforeProperties(JnlpConnectionState)} or
+     *                               {@link JnlpConnectionStateListener#afterProperties(JnlpConnectionState)}.
+     */
+    public void reject(ConnectionRefusalException reason) {
+        if (lifecycle.compareTo(State.AFTER_PROPERTIES) > 0) {
+            throw new IllegalStateException("Events cannot be rejected after approval/rejection");
+        }
+        lifecycle = State.REJECTED;
+        rejection = reason;
+    }
+
+    /**
+     * Retrieves the previously stashed state.
+     * @param clazz the expected class of the stashed state.
+     * @param <S> the expected class of the stashed state.
+     * @return the stashed state.
+     * @throws IllegalStateException if invoked before {@link #approve()}
+     * @see #setStash(ListenerState)
+     */
+    @CheckForNull
+    public <S extends ListenerState> S getStash(Class<S> clazz) {
+        if (lifecycle.compareTo(State.APPROVED) < 0) {
+            throw new IllegalStateException("The connection has not been approved yet");
+        }
+        return clazz.cast(stash);
+    }
+
+    /**
+     * Stores some listener specific state for later retrieval.
+     *
+     * @param stash the state to stash.
+     * @param <S>   the expected class of the stashed state.
+     * @throws IllegalStateException if invoked before {@link #approve()}
+     * @see #getStash(Class)
+     */
+    public <S extends ListenerState> void setStash(@CheckForNull S stash) {
+        if (lifecycle.compareTo(State.APPROVED) < 0) {
+            throw new IllegalStateException("The connection has not been approved yet");
+        }
+        this.stash = stash;
+    }
+
+    /**
+     * Encapsulates the common event dispatch logic.
+     *
+     * @param handler the logic to apply.
+     */
+    private void fire(EventHandler handler) {
+        Iterator<JnlpConnectionStateListener> iterator = listeners.iterator();
+        JnlpConnectionState.fireIterator.set(iterator);
+        try {
+            final State lifecycle = this.lifecycle;
+            while (iterator.hasNext()) {
+                JnlpConnectionStateListener current = iterator.next();
+                handler.invoke(current, this);
+                if (lifecycle != this.lifecycle) {
+                    // a listener has changed the state, thus they are the owner
+                    listeners.retainAll(Collections.singleton(current));
+                    return;
+                }
+            }
+        } finally {
+            JnlpConnectionState.fireIterator.remove();
+        }
+    }
+
+    /**
+     * Advances the connection state to indicate that a connection has been "secured" and the property exchange
+     * is about to take place.
+     *
+     * @throws ConnectionRefusalException if the connection has been refused.
+     */
+    /*package*/ void fireBeforeProperties() throws ConnectionRefusalException {
```

Comment:
NIT: fire.* maybe not a perfect method naming pattern since the methods not only trigger listeners, but also change the internal state of the object